src/main/kotlin/org/rust/cargo/toolchain/tools/Rustfmt.kt

```diff
@@ -29,49 +36,74 @@ fun RsToolchainBase.rustfmt(): Rustfmt = Rustfmt(this)
 class Rustfmt(toolchain: RsToolchainBase) : RustupComponent(NAME, toolchain) {
 
     fun reformatDocumentTextOrNull(cargoProject: CargoProject, document: Document): String? {
+        val project = cargoProject.project
         return createCommandLine(cargoProject, document)
-            ?.execute(cargoProject.project, stdIn = document.text.toByteArray())
-            ?.unwrapOrElse { if (isUnitTestMode) throw it else return null }
-            ?.stdout
+            ?.execute(project, stdIn = document.text.toByteArray())
+            ?.unwrapOrElse { e ->
+                e.showRustfmtError(project)
+                if (isUnitTestMode) throw e else return null
+            }?.stdout
     }
 
     fun createCommandLine(cargoProject: CargoProject, document: Document): GeneralCommandLine? {
         val file = document.virtualFile ?: return null
         if (file.isNotRustFile || !file.isValid) return null
 
-        val arguments = buildList<String> {
+        val project = cargoProject.project
+        val settings = project.rustfmtSettings
+        val arguments = mutableListOf<String>().apply {
+            if (settings.channel != RustChannel.DEFAULT) add("+${settings.channel}")
+            addAll(ParametersListUtil.parse(settings.additionalArguments))
+
+            removeAll { it.startsWith("--emit") }
             add("--emit=stdout")
 
-            val configPath = findConfigPathRecursively(file.parent, stopAt = cargoProject.workingDirectory)
-            if (configPath != null) {
-                add("--config-path")
-                add(configPath.toString())
+            if (none { it.startsWith("--config-path") }) {
+                val configPath = findConfigPathRecursively(file.parent, stopAt = cargoProject.workingDirectory)
+                if (configPath != null) {
+                    add("--config-path=$configPath")
+                }
             }
 
-            val currentRustcVersion = cargoProject.rustcInfo?.version?.semver
-            if (currentRustcVersion != null) {
-                val edition = runReadAction {
-                    val psiFile = file.toPsiFile(cargoProject.project)
-                    psiFile?.edition ?: Edition.DEFAULT
+            if (none { it.startsWith("--edition") }) {
+                val currentRustcVersion = cargoProject.rustcInfo?.version?.semver
+                if (currentRustcVersion != null) {
+                    val edition = runReadAction {
+                        val psiFile = file.toPsiFile(project)
+                        psiFile?.edition ?: Edition.DEFAULT
+                    }
+                    add("--edition=${edition.presentation}")
                 }
-                add("--edition=${edition.presentation}")
             }
```

Comment:
nit: Maybe introduce something like 
```
private fun MutableList<String>.addArgument(flagName: String, value: () -> String) {}
```
to extract common code for addition of some argument if it's not passed yet and use it for `config-path` and `edition`